streaming/base/util.py

```diff
@@ -196,3 +203,126 @@ def get_import_exception_message(package_name: str, extra_deps: str) -> str:
     return f'Streaming was installed without {package_name} support. ' + \
             f'To use {package_name} related packages with Streaming, run ' + \
             f'`pip install \'mosaicml-streaming[{package_name}]\'`.'
+
+
+def merge_index(folder_urls: List[Union[str, Tuple[str, str]]],
+                out: Union[str, Tuple[str, str]],
+                *,
+                keep_local: bool = True,
+                overwrite: bool = True,
+                download_timeout: int = 100) -> None:
+    """Merge index.json from a list of remote or local directories into one for streaming and save the merged index to local or remote.
+
+    Args:
+        folder_urls (Iterable): folders that contain index.json for the partition
+            each element can take the form of a single path string or a tuple string
+
+            for each  url in folder_urls, if url is
+                1. tuple (local, remote): check if local is accessible.
+                    -> Yes: use local index to merge
+                    -> No:  download from remote first, then merge
+                2. str (local path): use local path to merge.
+                    raise FileNotFoundError if any local index is not accessible
+                3. str (remote url): download to a temp directory first, then merge
+
+        out (Union[str, Tuple[str, str]]): path to put the merged index file
+        keep_local (bool): Keep local copy of the merged index file. Defaults to ``True``
+        overwrite (bool): Overwrite merged index file in out if there exists one.Defaults to ``True``
+        download_timeout (int): The allowed time for downloading each json file
+            defaults to 60, same as streaming.download_file
+    """
+    # Import here to avoid circular import error
+    from streaming.base.storage.upload import CloudUploader
+    from streaming.base.storage.download import download_file
+
+    if not folder_urls:
+        logger.warning('No partitions exist, no index merged')
+        return
+
+    # This is the index json file name, e.g., it is index.json as of 0.6.0
+    index_basename = get_index_basename()
+
+    if os.path.exists(os.path.join(out, index_basename)) and overwrite:
+        logger.warning('Merged index already exists. no index merged if overwrite=False')
+        return
+
+    # Prepare a temp folder to download index.json rom remote if necessary. Removed in the end.
+    temp_root = mkdtemp()
```

Comment:
You should use a construct like
`with tempfile.TemporaryDirectory() as tmp:` 
to ensure the temporary directory (and it's contents are always deleted even on exceptions)